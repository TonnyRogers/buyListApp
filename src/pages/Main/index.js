import React, {useState} from 'react';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {useNavigation} from '@react-navigation/native';
import {useDispatch, useSelector} from 'react-redux';

import {
  Container,
  Content,
  Title,
  ListButton,
  ListName,
  BuyListener,
  NewListButton,
  EmptyList,
  EmptyListText,
  ListInfo,
  ListIcon,
  SubmitButton,
} from './styles';

import {
  createListRequest,
  closeListModal,
  openListModal,
} from '../../store/modules/list/actions';
import Modal from '../../components/Modal';
import Input from '../../components/TextInput';

const Main = ({navigation}) => {
  // const navigation = useNavigation();
  const lists = useSelector((state) => state.list);
  const dispatch = useDispatch();
  const isNewListModalOpen = lists.creatListModalOpen;

  const [listName, setListName] = useState('');

  function listDetail(list) {
    navigation.navigate('ListDetail', {list});
  }

  function newListModalClose() {
    dispatch(closeListModal());
  }

  function newListModalOpen() {
    dispatch(openListModal());
  }

  function handleCreateList() {
    if (!listName) {
      return;
    }

    dispatch(createListRequest(listName));
    setListName('');
    newListModalClose();
  }

  return (
    <Container>
      <Content>
        <Title>Suas Listas de Compras</Title>
        <BuyListener
          testID="list-list"
          data={lists.data}
          keyExtractor={(item) => String(item.id)}
          renderItem={({item}) => (
            <ListButton
              testID={`list-${item.id}`}
              onPress={() => listDetail(item)}>
              <ListIcon name="shopping-cart" size={24} color="#FFF" />
              <ListInfo>
                <ListName>{item.name}</ListName>
              </ListInfo>
            </ListButton>
          )}
          ListEmptyComponent={() => (
            <EmptyList>
              <Icon name="remove-shopping-cart" size={24} color="#028090" />
              <EmptyListText>Você não possui listas</EmptyListText>
            </EmptyList>
          )}
        />
      </Content>

      <NewListButton testID="new-list-btn" onPress={newListModalOpen}>
        <Icon name="add-shopping-cart" size={24} color="#00a896" />
      </NewListButton>

      <Modal
        testID="new-list-modal"
        visible={isNewListModalOpen}
        onRequestClose={newListModalClose}
        title="Nova Lista">
        <Input
          testID="new-list-input"
          placeholder="Dê um nome da sua nova lista"
          value={listName}
          onChange={setListName}
        />
        <SubmitButton testID="new-list-btn-submit" onPress={handleCreateList}>
          <Icon name="check" size={24} color="#FFF" />
        </SubmitButton>
      </Modal>
    </Container>
  );
};

export default Main;
