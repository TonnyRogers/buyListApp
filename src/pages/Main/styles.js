import styled from 'styled-components/native';
import Icon from 'react-native-vector-icons/MaterialIcons';

// https://coolors.co/05668d-028090-00a896-02c39a-f0f3bd Pallete

export const Container = styled.View`
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
`;

export const Content = styled.View`
  margin: 50px 10px 10px;
  background: rgba(255, 255, 255, 0.5);
  flex: 1;
  padding: 12px;
`;

export const BuyListener = styled.FlatList``;

export const Title = styled.Text`
  font-size: 24px;
  text-align: center;
  margin-bottom: 20px;
  font-weight: 700;
`;

export const ListButton = styled.TouchableOpacity`
  flex-direction: row;
  align-items: center;
  background: #028090;
  margin: 5px 0;
  padding: 12px;
  border-radius: 4px;
`;

export const ListInfo = styled.View`
  border-left-width: 1px;
  border-left-color: #fff;
`;

export const ListIcon = styled(Icon)`
  margin-right: 8px;
`;

export const ListName = styled.Text.attrs({
  numberOfLines: 1,
})`
  font-size: 15px;
  font-weight: bold;
  color: #fff;
  margin-left: 8px;
`;

export const EmptyList = styled.View`
  flex-direction: row;
  align-items: center;
  margin: 5px 0;
  padding: 12px;
  border-radius: 4px;
  border: 1px dashed #028090;
`;

export const EmptyListText = styled.Text`
  font-size: 16px;
  color: #028090;
  margin-left: 8px;
`;

export const NewListButton = styled.TouchableOpacity`
  height: 48px;
  width: 48px;
  border-radius: 24px;
  background: rgba(255, 255, 255, 0.3);
  align-items: center;
  justify-content: center;
  align-self: flex-end;
  margin-bottom: 12px;
  margin-right: 12px;
`;

export const SubmitButton = styled.TouchableOpacity`
  height: 48px;
  width: 48px;
  border-radius: 24px;
  background: #028090;
  align-items: center;
  justify-content: center;
  align-self: center;
`;
