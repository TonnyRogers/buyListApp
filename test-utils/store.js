import {createStore, applyMiddleware} from 'redux';
import createSagaMiddleware from 'redux-saga';
import thunk from 'redux-thunk';
import rootReducer from '../src/store/modules/rootReducer';
import rootSaga from '../src/store/modules/rootSaga';

export function buildStore(initialState) {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares = [thunk, sagaMiddleware];

  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(...middlewares),
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
