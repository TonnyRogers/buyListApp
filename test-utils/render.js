import React from 'react';
import {Provider} from 'react-redux';
import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

import {buildStore} from './store';

export function renderWithRedux(component, {state} = {}) {
  const store = buildStore(state);
  const queries = render(
    <Provider store={store}>
      <NavigationContainer>{component}</NavigationContainer>
    </Provider>,
  );

  return {
    ...queries,
    store,
  };
}
