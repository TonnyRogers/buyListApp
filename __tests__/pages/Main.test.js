/* eslint-disable no-undef */
/* eslint-disable global-require */
import React from 'react';
import {fireEvent, cleanup} from '@testing-library/react-native';

import {renderWithRedux} from '../../test-utils/render';

import Main from '../../src/pages/Main';
// import MockedNavigator from '../../jest/MockedNavigator';

jest.mock(
  './path/to/the/image.png',
  () =>
    '/node_modules/@react-navigation/stack/lib/commonjs/views/assets/back-icon.png',
);

// const spyUseSelector = jest.spyOn(redux, 'useSelector');

// spyUseSelector.mockReturnValue({
//   list: {},
//   products: {},
// });

describe('Main page', () => {
  afterEach(cleanup);

  const initialState = {
    list: {
      data: [],
      loading: false,
      creatListModalOpen: false,
      selectedProducts: [],
    },
  };

  function renderMain(testState) {
    const state = {
      ...initialState,
      ...(testState || initialState),
    };

    return renderWithRedux(<Main />, {state});
  }

  it('open create new list modal', async () => {
    const rendered = renderMain();

    fireEvent.press(await rendered.findByTestId('new-list-btn'));

    renderMain({
      ...(initialState.list.creatListModalOpen = true),
    });

    expect(
      (await rendered.findByTestId('new-list-modal')).props.visible,
    ).toBeTruthy();
  });

  it('create new list', async () => {
    const rendered = renderMain();

    fireEvent.press(await rendered.findByTestId('new-list-btn'));

    renderMain({
      ...(initialState.list.creatListModalOpen = true),
    });

    fireEvent.changeText(
      await rendered.findByTestId('new-list-input'),
      'Presentes',
    );

    fireEvent.press(await rendered.findByTestId('new-list-btn-submit'));

    renderMain({
      ...(initialState.list.data = {id: 1, name: 'Presentes', total: 0.0}),
    });

    console.log(await rendered.findAllByText('Presentes'));
    expect(await rendered.findAllByText('Presentes')).toBeTruthy();
  });
});
