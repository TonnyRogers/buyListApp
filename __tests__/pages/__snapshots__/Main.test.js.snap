// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page open create new list modal 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0.5)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "marginBottom": 10,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 50,
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 24,
              "fontWeight": "700",
              "marginBottom": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Suas Listas de Compras
      </Text>
      <RCTScrollView
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        maxToRenderPerBatch={10}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Array [
            Object {},
          ]
        }
        testID="list-list"
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "#028090",
                  "borderRadius": 4,
                  "borderStyle": "dashed",
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "marginBottom": 5,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "marginTop": 5,
                  "paddingBottom": 12,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 12,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#028090",
                    "fontSize": 24,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#028090",
                    "fontSize": 16,
                    "marginLeft": 8,
                  },
                ]
              }
            >
              Você não possui listas
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      accessible={true}
      focusable={true}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-end",
          "backgroundColor": "rgba(255, 255, 255, 0.3)",
          "borderRadius": 24,
          "height": 48,
          "justifyContent": "center",
          "marginBottom": 12,
          "marginRight": 12,
          "opacity": 1,
          "width": 48,
        }
      }
      testID="new-list-btn"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#00a896",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      style={
        Array [
          Object {},
        ]
      }
      testID="new-list-modal"
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.6)",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
                "borderRadius": 4,
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "marginBottom": 20,
                "marginLeft": 10,
                "marginRight": 10,
                "marginTop": 20,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#e1e1e1",
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingBottom": 10,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              Nova Lista
            </Text>
            <View
              accessible={true}
              focusable={true}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-end",
                  "backgroundColor": "#028090",
                  "borderRadius": 24,
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": 30,
                }
              }
              testID="new-list-close-btn"
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#FFF",
                      "fontSize": 24,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            placeholder="Dê um nome da sua nova lista"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "backgroundColor": "#f1f1f1",
                  "borderColor": "#e1e1e1",
                  "borderRadius": 4,
                  "borderStyle": "solid",
                  "borderWidth": 1,
                  "fontSize": 14,
                  "height": 46,
                  "marginBottom": 10,
                  "marginLeft": 10,
                  "marginRight": 10,
                  "marginTop": 40,
                  "paddingLeft": 8,
                },
              ]
            }
            testID="new-list-input"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            accessible={true}
            focusable={true}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#028090",
                "borderRadius": 24,
                "height": 48,
                "justifyContent": "center",
                "opacity": 1,
                "width": 48,
              }
            }
            testID="new-list-btn-submit"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#FFF",
                    "fontSize": 24,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
